project/Core/Render/CoreCarbon.swift

```diff
@@ -1,68 +1,45 @@
 //  Copyright © 2019 The nef Authors.
 
-import Foundation
 import NefModels
+import Bow
+import BowEffects
 
-public protocol CarbonDownloader: class {
-    func carbon(withConfiguration configuration: CarbonModel, filename: String) -> Result<String, CarbonError>
+public protocol CarbonDownloader {
+    func carbon(configuration: CarbonModel) -> IO<CarbonError, Image>
 }
 
-public struct CarbonGenerator: InternalRender {
-    private let downloader: CarbonDownloader
-    private let style: CarbonStyle
-    private let output: String
 
-    public init(downloader: CarbonDownloader, style: CarbonStyle, output: String) {
-        self.downloader = downloader
-        self.style = style
-        self.output = output
-    }
-    
-    public func isValid(trace: String) -> Bool {
-        return !trace.contains("☓")
-    }
-    
-    internal func render(node: Node) -> String {
-        return node.carbon(downloader: self)
-    }
-}
-
-// MARK: - Node Downloader
-protocol CarbonCodeDownloader {
-    func carbon(code: String) -> String
-}
+extension NodeProcessor where D == CoreCarbonEnvironment, A == Image {
+    static var carbon: NodeProcessor {
+        func render(node: Node) -> EnvIO<D, CoreRenderError, A> {
+            EnvIO { env in
+                node.carbon(downloader: env.downloader, style: env.style)
+            }
+        }
 
-extension CarbonGenerator: CarbonCodeDownloader {
-    func carbon(code: String) -> String {
-        let configuration = CarbonModel(code: code, style: style)
-        let result = downloader.carbon(withConfiguration: configuration, filename: output)
-        
-        switch result {
-        case let .success(filename):
-            return "Downloading Carbon snippet for '\(filename)' ✓"
-            
-        case let .failure(carbonError):
-            return """
-                    Downloading Carbon snippet for '\(carbonError.filename)' ☓
-                        error: \(carbonError.cause)
-                        code snippet:
-                            \(carbonError.snippet)
-                   """
+        func merge(nodes: [A]) -> EnvIO<D, CoreRenderError, NEA<A>> {
+            let nodes = nodes.filter { image in !image.isEmpty }
+            guard !nodes.isEmpty else { return EnvIO.raiseError(.emptyNode)^ }
+            return EnvIO.pure(NEA(head: nodes[0], tail: Array(nodes[1...])))^
```

Comment:
I think we have a constructor that does exactly this: `NEA.fromArrayUnsafe` or something like that.

Reply:
sure, I will update it